tests/PHPStan/Analyser/data/filter_var.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace FilterVar;
+
+use function PHPStan\Testing\assertType;
+
+function doBar($mixed) {
+    filter_var($mixed, FILTER_VALIDATE_FLOAT);
+    assertType('mixed', $mixed);
+
+	if (filter_var($mixed, FILTER_VALIDATE_FLOAT)) {
+		assertType('float', $mixed);
+	}
+	assertType('mixed', $mixed);
+
+	if (filter_var($mixed, FILTER_VALIDATE_FLOAT) === true) {
+		assertType('float', $mixed);
+	}
+	assertType('mixed', $mixed);
+
+	if (filter_var($mixed, FILTER_VALIDATE_FLOAT) === false) {
+		assertType('mixed', $mixed); // could be mixed~float
+	}
+	assertType('mixed', $mixed);
+
+	if (filter_var($mixed, FILTER_VALIDATE_FLOAT)) {
+		assertType('float', $mixed);
+		return;
+	}
+	assertType('mixed', $mixed); // could be mixed~float
+}
+
+function doFoo($mixed) {
+	if (filter_var($mixed, FILTER_VALIDATE_FLOAT)) {
+		assertType('float', $mixed);
+	}
+	if (filter_var($mixed, FILTER_VALIDATE_INT)) {
+		assertType('int<min, -1>|int<1, max>', $mixed);
+	}
+	if (filter_var($mixed, FILTER_VALIDATE_BOOL)) {
+		assertType('0|1|bool|string', $mixed);
+	}
+	if (filter_var($mixed, FILTER_VALIDATE_BOOLEAN)) {
+		assertType('0|1|bool|string', $mixed);
```

Comment:
Looking at https://www.php.net/manual/en/filter.filters.validate.php this is a huge can of worms. I checked randomly behaviour of `FILTER_VALIDATE_BOOLEAN` and I think this assert is wrong here.

1) It cannot be `false`.
2) It can only be a few selected strings.

Also we can definitely say something about the values in falsy context too?

Reply:
the "few" strings are case-insensitive - listing them all might get a longer union. but still doable if you prefer.
my tests suggest it can get `false` though

see
- https://www.php.net/manual/en/function.filter-var.php#121263
- https://3v4l.org/AV57X


> Also we can definitely say something about the values in falsy context too?

its kind of a subtractable thing then, right? or do you have something more meaningful in mind?